daemon/service-db.cc

```diff
@@ -56,149 +74,277 @@ MLServiceDB::~MLServiceDB ()
-
-  leveldb_put (db_obj, db_woptions, (char *) hash, sizeof (std::size_t) * 2,
-      value.c_str (), value.size (), &err);
-  if (err != nullptr) {
-    g_warning
-        ("Failed to call leveldb_put () for the name, '%s' of the pipeline description (size: %zu bytes / description: '%.40s')",
-        name.c_str (), value.size (), value.c_str ());
-    g_warning ("leveldb_put () has returned an error: %s", err);
-    leveldb_free (err);
-    throw std::runtime_error ("Failed to put()!");
+
+  /* (key, description) */
+  sql = g_strdup_printf ("INSERT OR REPLACE INTO tblPipeline (key, description) VALUES ('%s', '%s');",
+      key_with_prefix.c_str (), description.c_str ());
+  rc = sqlite3_exec (_db, sql, nullptr, nullptr, &errmsg);
+  g_free (sql);
+  if (rc != SQLITE_OK) {
+    g_warning ("Failed to insert pipeline description with name %s: %s (%d)", name.c_str (), errmsg, rc);
+    sqlite3_clear_errmsg (errmsg);
+    throw std::runtime_error ("Failed to insert pipeline description.");
   }
 }
 
 /**
- * @brief Get the value with the given name.
+ * @brief Get the pipeline description with the given name.
  * @param[in] name The unique name to retrieve.
- * @param[out] value The pipeline corresponding with the given name.
+ * @param[out] description The pipeline corresponding with the given name.
  */
 void
-MLServiceDB::get (const std::string name, std::string & out_value)
+MLServiceDB::get_pipeline (const std::string name, std::string &description)
 {
-  char *err = nullptr;
+  int rc;
+  char *sql;
   char *value = nullptr;
-  gsize read_len;
+  sqlite3_stmt *res;
 
   if (name.empty())
     throw std::invalid_argument ("Invalid name parameters!");
 
   std::string key_with_prefix = DB_KEY_PREFIX;
   key_with_prefix += name;
-  std::size_t hash[2] = { std::hash<std::string>{}(key_with_prefix), 0U };
-
-  value = leveldb_get (db_obj, db_roptions, (char *) hash, sizeof (std::size_t) * 2,
-      &read_len, &err);
-  if (err != nullptr) {
-    g_warning
-        ("Failed to call leveldb_get() for the name %s. Error message is %s.",
-        name.c_str (), err);
-    leveldb_free (err);
-    if (value)
-      leveldb_free (value);
-    throw std::runtime_error ("Failed to get()!");
-  }
 
-  if (!value) {
-    g_warning
-        ("Failed to find the key %s. The key should be set before reading it.",
-        name.c_str ());
-    throw std::invalid_argument ("Failed to find the key.");
+  sql = g_strdup_printf ("SELECT description FROM tblPipeline WHERE key = '%s';",
+      key_with_prefix.c_str ());
+  rc = sqlite3_prepare_v2 (_db, sql, -1, &res, nullptr);
+  g_free (sql);
+  if (rc != SQLITE_OK) {
+    g_warning ("Failed to get pipeline description with name %s: %s (%d)", name.c_str (), sqlite3_errmsg (_db), rc);
+    goto error;
   }
 
-  out_value = std::string (value, read_len);
-  leveldb_free (value);
-  return;
+  rc = sqlite3_step (res);
+  if (rc == SQLITE_ROW)
+    value = g_strdup_printf ("%s", sqlite3_column_text (res, 0));
+
```

Comment:
How about add
 if ( rc != SQLITE_DONE || rc != SQLITE_ROW)
 {
      /* print err messsage and handling*/
      goto error;
 }

Reply:
It is a final step to fetch value from pipeline table and will be handled in line 215.

Reply:
Okay, I understand. thank you very much.